src/Pate/Verification/AbstractDomain.hs

```diff
@@ -0,0 +1,506 @@
+  W4.IsSymExprBuilder sym =>
+  sym ->
+  W4.SymExpr sym tp ->
+  AbsRange tp ->
+  IO (PS.AssumptionFrame sym)
+applyAbsRange sym e rng = case rng of
+  AbsIntConstant i -> do
+    i' <- W4.intLit sym i
+    return $ PS.exprBinding e i'
+  AbsBVConstant w bv -> do
+    bv' <- W4.bvLit sym w bv
+    return $ PS.exprBinding e bv'
+  AbsBoolConstant True -> return $ PS.exprBinding e (W4.truePred sym)
+  AbsBoolConstant False -> return $ PS.exprBinding e (W4.falsePred sym)
+  AbsUnconstrained{} -> return mempty
+
+absDomainValToAsm ::
+  W4.IsSymExprBuilder sym =>
+  sym ->
+  PSR.MacawRegEntry sym tp ->
+  MacawAbstractValue sym tp ->
+  IO (PS.AssumptionFrame sym)
+absDomainValToAsm sym e (MacawAbstractValue absVal) = case PSR.macawRegRepr e of
+  CLM.LLVMPointerRepr{} -> do
+    CLM.LLVMPointer region offset <- return $ PSR.macawRegValue e
+    (Ctx.Empty Ctx.:> regAbs Ctx.:> offsetAbs) <- return $ absVal
+    regionInt <- W4.natToInteger sym region
+    regFrame <- applyAbsRange sym regionInt regAbs
+    offFrame <- applyAbsRange sym offset offsetAbs
+    return $ regFrame <> offFrame
+  CT.BoolRepr -> do
+    b <- return $ PSR.macawRegValue e
+    (Ctx.Empty Ctx.:> bAbs) <- return $ absVal
+    applyAbsRange sym b bAbs
+  CT.StructRepr Ctx.Empty -> return mempty
+  _ -> error "applyAbsDomainVal: unexpected type for abstract domain"
+
+
+-- | Construct an 'PS.AssumptionFrame' asserting
+-- that the values in the given initial state
+-- ('PS.SimInput') are necessarily in the given abstract domain
+absDomainValsToAsm ::
+  forall sym arch bin.
+  W4.IsSymExprBuilder sym =>
+  MapF.OrdF (W4.SymExpr sym) =>
+  MC.RegisterInfo (MC.ArchReg arch) =>
+  sym ->
+  PS.SimState sym arch bin ->
+  AbstractDomainVals sym arch bin ->
+  IO (PS.AssumptionFrame sym)
+absDomainValsToAsm sym st vals = do
+  memFrame <- MapF.foldrMWithKey getCell mempty (absMemVals vals)
+  regFrame <- fmap PRt.collapse $ PRt.zipWithRegStatesM (PS.simRegs st) (absRegVals vals) $ \_ val absVal ->
+    Const <$> absDomainValToAsm sym val absVal
+  return $ memFrame <> regFrame
+  where
+    getCell ::
+      PMC.MemCell sym arch w ->
+      MemAbstractValue sym w ->
+      PS.AssumptionFrame sym ->
+      IO (PS.AssumptionFrame sym)
+    getCell cell (MemAbstractValue absVal) frame = do
+      val <- IO.liftIO $ PMC.readMemCell sym (PS.simMem st) cell
+      frame' <- absDomainValToAsm sym (PSR.ptrToEntry val) absVal
+      return $ frame <> frame'
+
+-- | Construct a 'W4.Pred' asserting
+-- that the values in the given initial state
+-- ('PS.SimInput') are necessarily in the given abstract domain
+absDomainValsToPred ::
+  forall sym arch bin.
+  W4.IsSymExprBuilder sym =>
+  MapF.OrdF (W4.SymExpr sym) =>
+  MC.RegisterInfo (MC.ArchReg arch) =>
+  sym ->
+  PS.SimState sym arch bin ->
+  AbstractDomainVals sym arch bin ->
+  IO (W4.Pred sym)
+absDomainValsToPred sym st vals = do
+  asm <- absDomainValsToAsm sym st vals
+  PS.getAssumedPred sym asm
+
+absDomainToPrecond ::
```

Comment:
Presumably this is meant to be assumed somewhere?

Reply:
Yes, I'll add comments.